worlds/saving_princess/Items.py

```diff
@@ -0,0 +1,98 @@
+from typing import Optional, Dict
+
+from BaseClasses import Item, ItemClassification as ItemClass
+
+from .Constants import *
+
+
+class SavingPrincessItem(Item):
+    game: str = GAME_NAME
+
+
+class ItemData:
+    item_class: ItemClass
+    code: Optional[int]
+    count: int  # Number of copies for the item that will be made of class item_class
+    count_extra: int  # Number of extra copies for the item that will be made as useful
+
+    def __init__(self, item_class: ItemClass, code: Optional[int] = None, count: int = 1, count_extra: int = 0):
+        self.item_class = item_class
+
+        self.code = code
+        if code is not None:
+            self.code += BASE_ID
+
+        # if this is filler, a trap or an event, ignore the count
+        if self.item_class == ItemClass.filler or self.item_class == ItemClass.trap or code is None:
+            self.count = 0
+            self.count_extra = 0
+        else:
+            self.count = count
+            self.count_extra = count_extra
+
+    def create_item(self, player: int):
```

Comment:
Instead of creating these every single time you create an item, why not just make lookup tables directly beforehand that are shared and then use those?

Reply:
Sorry, I'm not sure I follow.
I think you mean `count` and `count_extra`? But I'm pretty sure those are only touched once per entry in my item dictionary, which is the only place `ItemData` instances are made (unless I made a mistake somewhere).

Or maybe it's my class naming that is confusing? `ItemData` holds data used to create `SavingPrincessItem` instances, but the `SavingPrincessItem` class doesn't have `ItemData` in it.

Reply:
I'm saying that every time you call `create_item` the `index` and `name` variables get totally rebuilt. But since the item dict is static, you could just create these once, right after making the item dict itself instead of creating them many times

Reply:
I understand now. Maybe something like this would make more sense:
```py
    def create_item(self, player: int):
        return SavingPrincessItem(item_data_names[self], self.item_class, self.code, player)

[... all of the ItemData dict stuff ...]

item_data_names: Dict[ItemData, str] = {value: key for key, value in item_dict.items()}
````

Reply:
Looks good to me